Utils/azuretools-core/src/com/microsoft/azuretools/ui/embeddedbrowser/JxBrowserUtil.java

```diff
@@ -22,42 +22,303 @@
             browserView = browserViewConstructor.newInstance(browser);
 
             if (browserView != null) {
                 Method loadURLMethod = browserClass.getMethod("loadURL", String.class);
-                loadURLMethod.invoke(url);
+                loadURLMethod.invoke(browser, url);
             }
         } catch (Exception e) {
-            throw new Exception("Fail to load JxBrowser or load URL: " + e.getMessage());
+            throw new JxBrowserException("Fail to load JxBrowser or load URL: " + e.getMessage());
         }
 
         return (JComponent)browserView;
     }
+
+    /**
+     * Download JxBrowser jars (i.e. jxbrowser-win32-6.16.jar and license.jar) and check digest (with retries) into given targetPath
+     * @param filePathNames
+     * @return
+     */
+    private static boolean downloadFilesFromAzure(@NotNull List<Path> filePathNames) {
+        boolean result = true;
+        ExecutorService threadPool = Executors.newFixedThreadPool(2);
+        try {
+            downloadAndCheckDigest(threadPool, filePathNames);
+
+            threadPool.shutdown();
+            threadPool.awaitTermination((long) (MAX_WAIT_TIME_IN_SECONDS * Math.pow(RETRY_BACKOFF_FACTOR, MAX_RETRY_TIMES)), TimeUnit.SECONDS);
+        } catch (Exception ex) {
+            log.severe("Download JxBrowser jars failed with exception " + ex.getMessage());
+            result = false;
+        } finally {
+            threadPool.shutdownNow();
+        }
+
+        return result;
+    }
+
+    /**
+     * Check the file correctness by comparing the MD5 digest with the digest file. Return true if both file and digest exist and match.
+     * @param filePathName
+     * @param digestFilePathName
+     * @return
+     */
+    private static boolean checkFileDigest(@NotNull Path filePathName, @NotNull Path digestFilePathName) {
+        File targetFile = filePathName.toFile();
+        if (!targetFile.exists() || !targetFile.isFile()) {
+            log.warning(filePathName + " doesn't exist");
+            return false;
+        }
+
+        File digestFile = digestFilePathName.toFile();
+        if (!digestFile.exists() || !digestFile.isFile()) {
+            log.warning(digestFilePathName + " doesn't exist");
+            return false;
+        }
+
+        FileInputStream targetFileStream = null;
+        String digestFileContent = "content";
+        String digest = "digest";
+        boolean result = true;
+        try {
+            targetFileStream = new FileInputStream(targetFile);
+            digest = DigestUtils.md5Hex(targetFileStream).toUpperCase();
+            digestFileContent = FileUtils.readFileToString(digestFile, "UTF-8");
+        } catch (IOException e) {
+            log.warning(String.format("Fail to read file %s or digest %s", filePathName, digestFilePathName));
+            result = false;
+        } finally {
+            IOUtils.closeQuietly(targetFileStream);
+        }
```

Comment:
you can use try(targetFileStream = new FileInputStream(targetFile))  { } catch(ex) directly when closing the IOs is the only function in `finally`.

Reply:
Good suggestion. Actually add close there. Searched and found we are supposed to close before closeQuietly

---
In reply to: [156554704](https://github.com/Microsoft/azure-tools-for-java/pull/1296#discussion_r156554704) [](ancestors = 156554704)